virtualizarr/writers/icechunk.py

```diff
@@ -0,0 +1,179 @@
+from typing import TYPE_CHECKING, cast
+
+import numpy as np
+from xarray import Dataset
+from xarray.backends.zarr import encode_zarr_attr_value
+from xarray.core.variable import Variable
+from zarr import Group
+
+from virtualizarr.manifests import ChunkManifest, ManifestArray
+from virtualizarr.zarr import encode_dtype
+
+if TYPE_CHECKING:
+    from icechunk import IcechunkStore
+
+
+VALID_URI_PREFIXES = {
+    "s3://",
+    # "gs://",
+    # "azure://",
+    # "r2://",
+    # "cos://",
+    # "minio://",
+    "file:///",
+}
+
+
+def dataset_to_icechunk(ds: Dataset, store: "IcechunkStore") -> None:
+    """
+    Write an xarray dataset whose variables wrap ManifestArrays to an Icechunk store.
+
+    Currently requires all variables to be backed by ManifestArray objects.
+
+    Parameters
+    ----------
+    ds: xr.Dataset
+    store: IcechunkStore
+    """
+    from icechunk import IcechunkStore
+
+    if not isinstance(store, IcechunkStore):
+        raise TypeError(f"expected type IcechunkStore, but got type {type(store)}")
+
+    if not store.supports_writes:
+        raise ValueError("supplied store does not support writes")
+
+    # TODO only supports writing to the root group currently
+    # TODO pass zarr_format kwarg?
+    root_group = Group.from_store(store=store)
+
+    # TODO this is Frozen, the API for setting attributes must be something else
+    # root_group.attrs = ds.attrs
+    for k, v in ds.attrs.items():
+        root_group.attrs[k] = encode_zarr_attr_value(v)
+
+    return write_variables_to_icechunk_group(
+        ds.variables,
+        store=store,
+        group=root_group,
+    )
+
+
+def write_variables_to_icechunk_group(
+    variables,
+    store,
+    group,
+):
+    for name, var in variables.items():
+        write_variable_to_icechunk(
+            store=store,
+            group=group,
+            name=name,
+            var=var,
+        )
+
+
+def write_variable_to_icechunk(
+    store: "IcechunkStore",
+    group: Group,
+    name: str,
+    var: Variable,
+) -> None:
+    """Write a single (possibly virtual) variable into an icechunk store"""
+
+    if isinstance(var.data, ManifestArray):
+        write_virtual_variable_to_icechunk(
+            store=store,
+            group=group,
+            name=name,
+            var=var,
+        )
+    else:
+        # TODO is writing loadable_variables just normal xarray ds.to_zarr?
+        # raise NotImplementedError()
+        print("skipping non-virtual variable", name)
```

Comment:
I think I would rather that this raised. Ideally we would raise a single message with the list of all non-virtual variables, telling the user that they can drop those from this datasets and write them into icechunk separately (even as part of the same commit).

In fact, couldn't we just make this work fairly easily using something like

```python
def write_variables_to_icechunk_group(
    variables,
    store,
    group,
):
    virtual_variables = {name: var for var in variables if isinstance(var.data, ManifestArray)}

    for name, var in virtual_variables.items():
        write_virtual_variables_to_icechunk(
            store=store,
            group=group,
            name=name,
            var=var,
        )

    loadable_variables = {name: var for var in variables if name not in virtual_variables}
    
    ds = xr.Dataset(loadable_variables)
    ds.to_zarr(icechunkstore)
```

Reply:
hmmmm that might indeed work. I switched to pass because raise would mean my test datasets didnt work (i never considered i could drop!! duh!).